nixos/modules/services/misc/matrix-dendrite.nix

```diff
@@ -0,0 +1,208 @@
+              This is also the last part of your UserID.
+            '';
+          };
+          private_key = lib.mkOption {
+            type = lib.types.nullOr lib.types.path;
+            example = "${workingDir}/matrix_key.pem";
+            default = "${workingDir}/matrix_key.pem";
+            description = ''
+              The path to the signing private key file, used to sign
+              requests and events.
+            '';
+          };
+          trusted_third_party_id_servers = lib.mkOption {
+            type = lib.types.listOf lib.types.str;
+            example = [ "matrix.org" ];
+            default = [ "matrix.org" "vector.im" ];
+            description = ''
+              Lists of domains that the server will trust as identity
+              servers to verify third party identifiers such as phone
+              numbers and email addresses
+            '';
+          };
+          disable_federation = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = ''
+              Disables federation. Dendrite will not be able to make
+              any outbound HTTP requests to other servers and the federation
+              API will not be exposed.
+            '';
+          };
+        };
+        options.client_api = {
+          registration_disabled = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = ''
+              Whether to disable user registration to the server
+              without the shared secret.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for dendrite, see:
+        <link xlink:href="https://github.com/matrix-org/dendrite/blob/master/dendrite-config.yaml"/>
+        for available options with which to populate settings.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.matrix-dendrite.tlsCert = lib.mkIf cfg.generateTls "/var/lib/matrix-dendrite/server.cert";
+    services.matrix-dendrite.tlsKey = lib.mkIf cfg.generateTls "/var/lib/matrix-dendrite/server.key";
+
+    services.matrix-dendrite.settings.client_api.registration_shared_secret = lib.mkIf (cfg.registrationSecretFile != null) "$REGISTRATION_SHARED_SECRET";
+
+    systemd.services.matrix-dendrite = {
+      description = "Dendrite Matrix homeserver";
+      after = [
+        "network.target"
+      ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "simple";
+        DynamicUser = true;
+        StateDirectory = "matrix-dendrite";
+        WorkingDirectory = workingDir;
+        ExecReload = "kill -HUP $MAINPID";
+        Restart = "on-failure";
+      };
+      script = lib.concatStringsSep "\n" (
+        lib.optionals (cfg.registrationSecretFile != null) [
+          "eval export $(cat ${cfg.registrationSecretFile} | ${pkgs.jq}/bin/jq -aR . )"
```

Comment:
In this case `jq` is not really helping because it was intended for escaping a single string literal, not multiple `KEY=VALUE` pairs. Anyway, I think it wasn't a big deal in the first place as it's only a problem if the secret starts with `"` or `'` which is extremely rare and can be manually escaped by the user. I only mentioned it because it looked like the review was stuck around escaping and this ensured that everything is escaped properly for a single string literal.

If you wanna go with the generic secrets file approach, then I would just remove `jq` to make things simpler as `envsubst` already solves most of the problems with `sed`.

Reply:
One more thing. If you're gonna use a generic env file approach, you could just use  `serviceConfig.EnvironmentFile` so you don't have to eval/export.

Reply:
That's exactly what I ended up doing :) Do you want to take a look at the latest changes? I believe this now (mostly) matches the mutter implementation.